data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BufferAccumulator.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.source;
+
+import com.amazon.dataprepper.model.buffer.Buffer;
+import com.amazon.dataprepper.model.record.Record;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Accumulates {@link Record} objects before placing them into a Data Prepper
+ * {@link Buffer}. This class is not thread-safe and should only be used by one
+ * thread at a time.
+ *
+ * @param <T> Type of record to accumulate
+ */
+class BufferAccumulator<T extends Record<?>> {
+    private final Buffer<T> buffer;
+    private final int recordsToAccumulate;
+    private final int timeoutMillis;
+
+    private final Collection<T> recordsAccumulated;
+
+    private BufferAccumulator(final Buffer<T> buffer, final int recordsToAccumulate, final int timeoutMillis) {
+        this.buffer = buffer;
+        this.recordsToAccumulate = recordsToAccumulate;
```

Comment:
Should this be validated as > 0?

`recordsToAccumulate` implies it's a list of records. Maybe `numberOfRecordsToAccumulate` or `recordAccumulationSize`?